internal/nix/search.go

```diff
@@ -106,3 +104,85 @@ func searchSystem(url, system string) (map[string]*Info, error) {
 	}
 	return parseSearchResults(out), nil
 }
+
+// searchSystemCache is a machine-wide cache of search results. It is shared by all
+// Devbox projects on the current machine. It is stored in the XDG cache directory.
+type searchSystemCache struct {
+	QueryToInfo map[string]map[string]*Info `json:"query_to_info"`
+}
+
+const (
+	// searchSystemCacheSubDir is a sub-directory of the XDG cache directory
+	searchSystemCacheSubDir   = "devbox/nix"
+	searchSystemCacheFileName = "search-system-cache.json"
+)
+
+var cache = searchSystemCache{}
+
+// SearchNixpkgsAttribute is a wrapper around searchSystem that caches results.
+// NOTE: we should be very conservative in where we use this function. `nix search`
+// accepts generalized `installable regex` as arguments but is slow. For certain
+// queries of the form `nixpkgs/<commit-hash>#attribute`, we can know for sure that
+// once `nix search` returns a valid result, it will always be the very same result.
+// Hence we can cache it locally and answer future queries fast, by not calling `nix search`.
+func SearchNixpkgsAttribute(query string) (map[string]*Info, error) {
```

Comment:
Why not just `CachedSearch`

Reply:
Could we throw in a regex to verify that the input query is `^github:NixOS/nixpkgs/[0-9a-f]{40}#`? Just to make sure we don't end up caching a ref like `github:NixOS/nixpkgs/nixpkgs-unstable#go_1_19`.

Reply:
good call, done